core/src/main/java/io/undertow/io/UndertowOutputStream.java

```diff
@@ -328,10 +329,14 @@ public void close() throws IOException {
         try {
             state |= FLAG_CLOSED;
             if (anyAreClear(state, FLAG_WRITE_STARTED) && channel == null) {
-                if (buffer == null) {
-                    exchange.getResponseHeaders().put(Headers.CONTENT_LENGTH, "0");
-                } else {
-                    exchange.getResponseHeaders().put(Headers.CONTENT_LENGTH, "" + buffer.position());
+                // Head requests may not write to the response, setting content-length directly instead
+                if (!(Methods.HEAD.equals(exchange.getRequestMethod())
+                        && exchange.getResponseHeaders().contains(Headers.CONTENT_LENGTH))) {
```

Comment:
Alternatively we could use this path any time `Content-Length` is already set regardless of the method, although that could impact existing code if values are set incorrectly. What do you think?

Reply:
It looks like now this condition duplicates the isHeadRequestWithContentLength method.

Reply:
🤦 thanks! Looks like I pushed too soon!